### PR TITLE
Suggester rebuild fixes

### PR DESCRIPTION
**Pull request by vladak**
This set of changes lowers the impact of suggester rebuild on the webapp/system.

### CODE REVIEW

**Comment by vladak** [comment]
The fix for #2593 was tested using newly added endpoint:
```
 curl -X PUT http://localhost:8080/source/api/v1/suggest/rebuild
```
while running:
```python
uri = 'http://localhost:8080/source/api/v1/search?full=fubar&projects=linux-master'

# Make sure the proxies are never used.
os.environ['http_proxy'] = ''
os.environ['https_proxy'] = ''

while True:
    try:
        try:
            rv = requests.get(uri, data='', headers=headers, proxies=None)
        except Exception as e:
            print(e)
            sys.exit(1)

        if rv.status_code != 200:
            print("Status code: " + str(rv.status_code))
            sys.exit(1)

        # pdb.set_trace()
        try:
            j = rv.json()
            print(j.get('time'))
        except Exception as e:
            print(e)
            sys.exit(1)
    except KeyboardInterrupt:
        sys.exit(0)
```
and observing that the requests were not stalled while the rebuild was in progress.

**Comment by vladak** [comment]
https://github.com/oracle/opengrok/wiki/Suggester will be updated once this is merged in.

**Comment by vladak** [comment]
Also, I had to fix some serial version UID's that IDEA suddenly started complaining about.

**Comment by vladak** [comment]
I had to fix the order of tests in `SuggesterControllerTest` since the newly added tests that call `rebuild()` impact other tests.

File: dev/checkstyle/suppressions.xml [hunk 1]
```diff
@@ -13,7 +13,7 @@
 
     <suppress checks="ParameterNumber" files="CtagsReader\.java|Definitions\.java|PerlLexHelper\.java|
         |JFlexXrefUtils\.java|RubyLexHelper\.java|FileAnalyzerFactory\.java|SearchController\.java|
-        |Context\.java|HistoryContext\.java|Indexer\.java" />
+        |Context\.java|HistoryContext\.java|Indexer\.java|Suggester\.java" />
```
Comment (Orviss) [inline dev/checkstyle/suppressions.xml]: I would avoid doing this. In no time we will have telescoping constructor anti-pattern.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/document/MandocException.java [hunk 2]
```diff
@@ -26,5 +26,7 @@
  * Represents an exception that occurs when mandoc exits non-zero.
  */
 public class MandocException extends Exception {
```
Comment (Orviss) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/document/MandocException.java]: Please add space between class and the attribute.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/index/Filter.java [hunk 3]
```diff
@@ -216,6 +216,8 @@ private Pattern compilePattern(String pattern) {
     }
 
     public static class PatternList extends ArrayList<String> {
+        private static final long serialVersionUID = -6883390970972775838L;
```
Comment (Orviss) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/index/Filter.java]: Missing space between class and attribute.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/index/IgnoredDirs.java [hunk 4]
```diff
@@ -32,6 +32,8 @@
  * @author Chandan
  */
 public final class IgnoredDirs extends Filter {
+    private static final long serialVersionUID = -6771850658877396699L;
```
Comment (Orviss) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/index/IgnoredDirs.java]: Missing space between class and attribute.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/index/IgnoredFiles.java [hunk 5]
```diff
@@ -32,6 +32,8 @@
  * @author Chandan
  */
 public final class IgnoredFiles extends Filter {
+    private static final long serialVersionUID = 3673125546519046976L;
```
Comment (Orviss) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/index/IgnoredFiles.java]: Missing space between class and attribute.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/index/IndexVersion.java [hunk 6]
```diff
@@ -50,6 +50,7 @@
      * exception thrown when index version does not match Lucene version
      */
     public static class IndexVersionException extends Exception {
+        private static final long serialVersionUID = -756788782277552544L;
```
Comment (Orviss) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/index/IndexVersion.java]: Missing space between class and attribute.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/index/IndexerException.java [hunk 7]
```diff
@@ -27,6 +27,7 @@
  * @author Jan Berg
  */
 public class IndexerException extends Exception {
+   private static final long serialVersionUID = -3009093549932264215L;
```
Comment (Orviss) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/index/IndexerException.java]: Missing space between class and attribute.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/ForbiddenSymlinkException.java [hunk 8]
```diff
@@ -28,5 +28,7 @@
  * encountered while evaluating a file system path.
  */
 public class ForbiddenSymlinkException extends Exception {
+      private static final long serialVersionUID = -5483867987230977680L;
```
Comment (Orviss) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/util/ForbiddenSymlinkException.java]: Missing space between class and attribute.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/web/messages/Message.java [hunk 9]
```diff
@@ -143,6 +143,7 @@ public int compareTo(final Message o) {
     }
 
     private static class DurationSerializer extends StdSerializer<Duration> {
+        private static final long serialVersionUID = 5275434375701446542L;
```
Comment (Orviss) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/web/messages/Message.java]: Missing space between class and attribute.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/web/messages/Message.java [hunk 10]
```diff
@@ -167,6 +168,7 @@ public void serialize(
     }
 
     private static class DurationDeserializer extends StdDeserializer<Duration> {
+        private static final long serialVersionUID = 5513386447457242809L;
```
Comment (Orviss) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/web/messages/Message.java]: Missing space between class and attribute.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/condition/DeliberateRuntimeException.java [hunk 11]
```diff
@@ -29,6 +29,8 @@
  * https://stackoverflow.com/questions/1754315/how-to-create-custom-exceptions-in-java
  */
 public class DeliberateRuntimeException extends RuntimeException {
+      private static final long serialVersionUID = 6492260220094558958L;
```
Comment (Orviss) [inline opengrok-indexer/src/test/java/org/opengrok/indexer/condition/DeliberateRuntimeException.java]: Missing space between class and attribute.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/SuggesterConfigTest.java [hunk 12]
```diff
@@ -0,0 +1,22 @@
+package org.opengrok.indexer.configuration;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class SuggesterConfigTest {
+    // TODO add more tests
```
Comment (Orviss) [inline opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/SuggesterConfigTest.java]: Please avoid using `TODO` comments, either remove the `TODO` comment or add the tests.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/SuggesterConfigTest.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
+package org.opengrok.indexer.configuration;
```
Comment (Orviss) [inline opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/SuggesterConfigTest.java]: Missing license?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/index/Filter.java (hunk 3 above)
Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/index/Filter.java]: is there a way how to enforce this in style checker ?

File: dev/checkstyle/suppressions.xml (hunk 1 above)
Comment (vladak) [reply dev/checkstyle/suppressions.xml]: filed #2610 to track them all. I'd like to get this in without introducing additional semi-complex changes.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/SuggesterConfigTest.java (hunk 12 above)
Comment (vladak) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/SuggesterConfigTest.java]: Right. Filed #2611 to track this.

File: opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/SuggesterConfigTest.java (hunk 13 above)
Comment (vladak) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/SuggesterConfigTest.java]: yep. Strange this was not caught by any check during build.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/index/Filter.java (hunk 3 above)
Comment (Orviss) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/index/Filter.java]: > is there a way how to enforce this in style checker ?

I'm not sure. Maybe somehow using regular expressions?

**Pull request merged**